### PR TITLE
release tool for lemming and operator

### PR DESCRIPTION
**Pull request by DanG100**

### CODE REVIEW

**Review by wenovus — COMMENTED** [review]
Another way to deal with having two projects in one is to use Git submodules: https://git-scm.com/book/en/v2/Git-Tools-Submodules

However I'm not convinced ours is the right use case, so this change overall looks good to me.

File: cmd/release/release.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+			sha, err := validateWorkDir()
+			if err != nil {
+				return err
+			}
+			fmt.Println("Running prerelease tests")
+			if err := triggerBuild(cmd.Context(), "operator-test", sha, false); err != nil {
+				return err
+			}
+
+			tag := fmt.Sprintf("operator/%s", args[0])
```
[the remark this answers is not in the document]
Comment (DanG100) [reply cmd/release/release.go:56]: yeah, it's how you version modules if when the go.mod isn't in the root of the repo https://go.dev/doc/modules/managing-source

File: cmd/release/release.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	if tag {
+		src = &cloudbuildpb.RepoSource{
+			Revision: &cloudbuildpb.RepoSource_TagName{
+				TagName: tagOrSHA,
+			},
+		}
+	}
+
+	op, err := c.RunBuildTrigger(ctx, &cloudbuildpb.RunBuildTriggerRequest{
+		Name:   fmt.Sprintf("projects/openconfig-lemming/locations/us-central1/triggers/%s", trigger),
```
[the remark this answers is not in the document]
Comment (DanG100) [reply cmd/release/release.go:131]: done

File: cmd/release/release.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+	}
+	if out, err := exec.Command("git", "push", "origin", tag).CombinedOutput(); err != nil {
+		return fmt.Errorf("failed to push tag: out %s, error %v", string(out), err)
+	}
+	return nil
+}
+
+// triggerBuild runs a cloud build trigger at the given tag if set, or the main branch if unset.
+func triggerBuild(ctx context.Context, trigger, tagOrSHA string, tag bool) error {
+	c, err := cloudbuild.NewClient(ctx, option.WithEndpoint("us-central1-cloudbuild.googleapis.com:443"))
```
[the remark this answers is not in the document]
Comment (DanG100) [reply cmd/release/release.go]: it was a workaround for a bug, liinked

File: cloudbuild/lemming-test.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    env:
+      - USERNAME=user
+      - SSH_ARGS=--internal-ip --ssh-key-expire-after=1d
+      - INSTANCE_NAME=kne-presubmit-$BUILD_ID
+      - INSTANCE_ARGS=--network cloudbuild-workers --image-project gep-kne --image-family kne --machine-type e2-standard-16 --preemptible
+      - ZONE=us-central1-a
+      - REMOTE_WORKSPACE=/tmp/workspace
+      - COMMAND=source /tmp/workspace/cloudbuild/lemming-test.sh 2>&1
+
+timeout: 1800s
```
[the remark this answers is not in the document]
Comment (DanG100) [reply cloudbuild/lemming-test.yaml:13]: added todo, not sure what a reasonable timeout is yet

**Comment by DanG100** [comment]
> Another way to deal with having two projects in one is to use Git submodules: https://git-scm.com/book/en/v2/Git-Tools-Submodules
> 
> However I'm not convinced ours is the right use case, so this change overall looks good to me.

that would require two repos, which i wanted to avoid. the operator won't be changing often, so i think the slight burden of the two go.mod is ok.

File: cloudbuild/operator-release.yaml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+steps:
+- name: 'gcr.io/cloud-builders/docker'
+  args: ['build', '-t', 'us-west1-docker.pkg.dev/${PROJECT_ID/release/operator:${_IMAGE_TAG}', '.', '-f', 'Dockerfile.operator']
+images: ['us-west1-docker.pkg.dev/${PROJECT_ID/release/operator:${_IMAGE_TAG}']
+substitutions:
+  _IMAGE_TAG: latest #default value
```
[the remark this answers is not in the document]
Comment (wenovus) [reply cloudbuild/operator-release.yaml:6]: Is `${TAG_NAME#/}` right? Should it be `${TAG_NAME#operator/}`

[the remark this answers is not in the document]
Comment (DanG100) [reply cloudbuild/operator-release.yaml:6]: ${TAG_NAME#/} means remove everything up to the 1st `/` which will do the same as `${TAG_NAME#operator/}`

[the remark this answers is not in the document]
Comment (wenovus) [reply cloudbuild/operator-release.yaml:6]: That's interesting, I'm seeing this on my local bash:
```bash
$ TAG_NAME=operator/2.2.2
$ echo ${TAG_NAME#/}
operator/2.2.2
```
Do you see something different?

**Review by wenovus — APPROVED** [review]

**Pull request merged**